lib/coarse_op.cuh

```diff
@@ -624,58 +172,287 @@ namespace quda {
+#if defined(WILSONCOARSE)
+        launch_host<compute_coarse_clover>(tp, stream, arg);
+#else
+        errorQuda("compute_coarse_clover not enabled for non-Wilson coarsenings");
+#endif
+      } else if (type == COMPUTE_REVERSE_Y) {
+        launch_host<reverse>(tp, stream, arg);
+      } else if (type == COMPUTE_DIAGONAL) {
+#if defined(WILSONCOARSE) || defined(COARSECOARSE)
+        launch_host<add_coarse_diagonal>(tp, stream, arg);
+#else
+        errorQuda("add_coarse_diagonal not enabled for staggered coarsenings");
+#endif
+      } else if (type == COMPUTE_STAGGEREDMASS) {
+#if defined(STAGGEREDCOARSE)
+        launch_host<add_coarse_staggered_mass>(tp, stream, arg);
+#else
+        errorQuda("add_coarse_staggered_mass not enabled for non-staggered coarsenings");
 #endif
+      } else if (type == COMPUTE_TMDIAGONAL) {
+#if defined(WILSONCOARSE) || defined(COARSECOARSE)
+        launch_host<add_coarse_tm>(tp, stream, arg);
+#else
+        errorQuda("add_coarse_tm not enabled for non-wilson coarsenings");
+#endif
+      } else if (type == COMPUTE_CONVERT) {
+        launch_host<convert>(tp, stream, arg);
+      } else if (type == COMPUTE_RESCALE) {
+        launch_host<rescale>(tp, stream, arg);
+      } else {
+        errorQuda("Undefined compute type %d", type);
+      }
+
+      if (compute_max) {
+        double max = *arg.max_h;
+        comm_allreduce_max(&max);
+        *arg.max_h = max;
+      }
+    }
+
+    /**
+       @brief Launcher for GPU instantiations of coarse-link construction
+    */
+    template <QudaFieldLocation location_> std::enable_if_t<location_ == QUDA_CUDA_FIELD_LOCATION>
+    Launch(Arg &arg, TuneParam &tp, ComputeType type, const qudaStream_t &stream)
+    {
+      if (compute_max) {
+        if (!activeTuning()) qudaMemsetAsync(arg.max_d, 0, sizeof(typename Arg::Float), stream);
+        arg.max = arg.max_d;
+      }
+
+      if (type == COMPUTE_UV) {
+
+        IF_CONSTEXPR (use_mma) {
+          if (compute_max) mma::launch_compute_uv_kernel(tp, ArgMax<Arg>(arg), arg.fineVolumeCB, stream, *this);
+          else mma::launch_compute_uv_kernel(tp, arg, arg.fineVolumeCB, stream, *this);
+        } else {
+          if (compute_max) launch_device<compute_uv>(tp, stream, ArgMax<Arg>(arg));
+          else launch_device<compute_uv>(tp, stream, arg);
+        }
+
+      } else if (type == COMPUTE_AV) {
+
+        if (from_coarse) errorQuda("compute_av should only be called from the fine grid");
+#if defined(GPU_CLOVER_DIRAC) && defined(WILSONCOARSE)
+        if (compute_max) launch_device<compute_av>(tp, stream, ArgMax<Arg>(arg));
+        else launch_device<compute_av>(tp, stream, arg);
+#else
+        errorQuda("Clover dslash has not been built");
+#endif
+
+      } else if (type == COMPUTE_TMAV) {
+
+        if (from_coarse) errorQuda("compute_tmav should only be called from the fine grid");
+#if defined(GPU_TWISTED_MASS_DIRAC) && defined(WILSONCOARSE)
+        launch_device<compute_tmav>(tp, stream, arg);
```

Comment:
To be thorough, should we add a `max` version here?

Reply:
We don't need a max variant of this, since we can analytically derive the bound.